lib/azure/armrest/environment.rb

```diff
@@ -74,15 +74,49 @@ def initialize(options)
           instance_variable_set("@#{key}", value)
         end
 
-        [:name, :active_directory_authority, :resource_manager_url].each do |key|
+        %i[name active_directory_authority resource_manager_url].each do |key|
           unless instance_variable_get("@#{key}")
             raise ArgumentError, "Mandatory argument '#{key}' not set"
           end
         end
       end
 
       alias authority_url active_directory_authority
+      alias login_endpoint active_directory_authority
       alias resource_url resource_manager_url
+      alias gallery_endpoint gallery_url
+      alias graph_endpoint graph_url
+
+      # Automatically discover and create an Environment given a resource
+      # manager endpoint. The +options+ hash must include an endpoint :url key.
+      # It may also include a :name (recommended), and http proxy options.
+      #
+      def self.discover(options = {})
+        url  = options.fetch(:url)
+        name = options[:name] || 'Custom'
+
+        uri = Addressable::URI.join(url, '/metadata/', 'endpoints')
+        uri.query = "api-version=1.0"
```

Comment:
why hardcoded `1.0`?

Reply:
That's what the docs say to use. As far as I could tell, metadata information cannot be gathered from the resource provider service, and that would require authentication first anyway.